### PR TITLE
[feat] Add utility function to look for available environment modules

### LINKED ISSUE #1159: Parametrizing tests based on module discovery
It would be a really great feature if ReFrame could parametrize tests based on the available modules. For example, something like this : 

```
@rfm.parameterized_test(*utility.all_modules('gromacs',min_version,max_version))
```

### PR DESCRIPTION
**Pull request by victorusu**
~~This PR extends the environment modules interface to include the `module avail` command.~~

~~This is the first step to allow the detection of modules to be included in the tests.~~

This PR adds the `find_modules()` utility function, which will retrieve all the available modules given a substring of a module name. Internally, it delegates to `module avail substr` of the current system's modules system, but it does so per system partition and respecting the ReFrame environments defined for each partition. It returns an iterator over a "set" of tuples of `(partition, environment, module)`. This works nicely with hierarchical modules, but it can also work fine with flat modules scheme. In this case, you can pass a dictionary to the `environ_mapping` argument, which maps module names to ReFrame environments. This way, `find_module()` will skip any combinations of `(partition, environment, module)` that are not meaningful. You can use the return value of this function in parameterized tests as follows, in order to generate automatically tests for different modules. Here is an example:

```python
@rfm.parameterized_test(*find_modules('netcdf'))
class MyTest(rfm.RegressionTest):
    def __init__(self, s, e, m):
        self.descr = f'{s}, {e}, {m}'
        self.valid_systems = [s]
        self.valid_prog_environs = [e]
        self.modules = [m]
```

For flat schemes you can use it as follows:

```python
my_find_modules = functools.partial(find_modules, environ_mapping={
    r'.*CrayGNU.*': {'PrgEnv-gnu'},
    r'.*CrayIntel.*': {'PrgEnv-intel'},
    r'.*CrayCCE.*': {'PrgEnv-cray'}
})

@rfm.parameterized_test(*my_find_modules('GROMACS'))
class MyTest(rfm.RegressionTest):
    def __init__(self, s, e, m):
        self.descr = f'{s}, {e}, {m}'
        self.valid_systems = [s]
        self.valid_prog_environs = [e]
        self.modules = [m]
```

Fixes #1159.

Still todo:

- [x] Document the function and add a tutorial example
- [x] Better type checking of the `toolchain_mapping` argument

Future work:

`find_modules()` can be somewhat slow since it calls `module avail` for each partition/environment combination. We could in the future implement a caching mechanism on top of it.

### CODE REVIEW

**Comment by victorusu** [comment]
@jenkins-cscs retry dom

File: reframe/core/modules.py [hunk 1]
```diff
@@ -152,6 +152,13 @@ def resolve_module(self, name):
     def backend(self):
         return(self._backend)
 
+    def available_modules(self):
```
Comment (vkarak) [inline reframe/core/modules.py:155]: The interface should accept a module name.

File: reframe/core/modules.py [hunk 2]
```diff
@@ -470,6 +484,17 @@ def _exec_module_command(self, *args, msg=None):
...
         exec(completed.stdout)
 
+    def available_modules(self):
+        avail = []
+        completed = self._run_module_command(
+            'available', '-t', msg="could not run module available")
+
+        for line in re.finditer(r'\S+[^:]$', completed.stderr, re.MULTILINE):
+            module = re.sub(r"\(default\)", "", line.group(0))
+            avail += [Module(module)]
```
Comment (vkarak) [inline reframe/core/modules.py:494]:
```suggestion
            module = re.sub(r'\(default\)', '', line)
            avail.append(Module(module))
```

File: reframe/core/modules.py [hunk 3]
```diff
@@ -470,6 +484,17 @@ def _exec_module_command(self, *args, msg=None):
         completed = self._run_module_command(*args, msg=msg)
         exec(completed.stdout)
 
+    def available_modules(self):
+        avail = []
+        completed = self._run_module_command(
+            'available', '-t', msg="could not run module available")
```
Comment (vkarak) [inline reframe/core/modules.py:490]:
```suggestion
            'avail', '-t', msg="could not run 'module avail'")
```

File: reframe/core/modules.py [hunk 4]
```diff
@@ -681,6 +706,17 @@ def name(self):
     def _module_command_failed(self, completed):
         return completed.stdout.strip() == 'false'
 
+    def available_modules(self):
+        avail = []
+        completed = self._run_module_command(
+            '-t', 'available', msg="could not run module available")
```
Comment (vkarak) [inline reframe/core/modules.py:712]: Why don't you use spider for Lmod?

Comment (victorusu) [reply reframe/core/modules.py:712]: Because spider can be cached. But that's a great idea too.
Do you want me to use it?

Comment (ekouts) [inline reframe/core/modules.py:712]:
```suggestion
            '-t', 'available', msg="could not run 'module available'")
```

File: reframe/core/modules.py [hunk 5]
```diff
@@ -681,6 +706,17 @@ def name(self):
...
+
+        for line in re.finditer(r'\S+[^:/]$', completed.stderr, re.MULTILINE):
+            module = re.sub(r"\(\S+\)", "", line.group(0))
```
Comment (ekouts) [inline reframe/core/modules.py:715]:
```suggestion
            module = re.sub(r'\(\S+\)', '', line.group(0))
```

**Comment by vkarak** [comment]
@mboisson @davidedelvento

This PR should do pretty much what we have been discussing in #1159. Could you test it?

**Comment by mboisson** [comment]
I'll have to put this on the pile. I haven't had time to touch our Reframe tests in months...

**Comment by vkarak** [comment]
@jenkins-cscs retry all

**Comment by davidedelvento** [comment]
I've been very busy with other stuff too, but trying various aspects of ReFrame in NCAR's environment is at the top of my todo list, just after a couple of other things which are due by the end of the month. So I should get to it in early October, and perhaps sooner. 
@vkarak  Do you need/want feedback earlier that this?

**Comment by vkarak** [comment]
@mboisson @davidedelvento I don't want to rush you, I just wanted to put that in your radar and test it whenever you have time. My plan is to merge this in the next weeks and ship it with ReFrame 3.2, but you can test it afterwards and we enhance it further.

**Comment by vkarak** [comment]
@jenkins-cscs retry all

**Review by vkarak — APPROVED** [review]

File: unittests/test_utility.py [hunk 6]
```diff
@@ -1293,3 +1297,73 @@ def test_cray_cle_info_missing_parts(tmp_path):
...
+        yield from temp_runtime(config_file, system,
+                                {'systems/modules_system': request.param})
+    except ConfigError as e:
+        pytest.skip(str(e))
+
+
+@pytest.fixture
+def modules_system(user_exec_ctx, monkeypatch):
+    # Pretend to be on a clean modules environment
+    #    monkeypatch.setenv('MODULEPATH', '')
```
Comment (vkarak) [inline unittests/test_utility.py:1329]: I should remove this.

**Comment by victorusu** [comment]
`lgtm`

**Pull request merged**